### PR TITLE
Frontend asset loading

### LINKED ISSUE #38: Create an example plugin hook for base.html template to allow registration and injection of frontend code

### LINKED ISSUE #41: Simply frontend_assets syntax
Current syntax looks like this:

```
  {% frontend_assets 'kolibri_core' plugin='kolibri.core' %}
```

It would be nice if we made it match python import syntax, so it looks like this:

```
  {% frontend_assets 'kolibri.core.KolibriCoreFrontend' %}
```

That should make the `name` attribute for `KolibriFrontendPluginBase` unnecessary.

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary

This is a WIP to implement and provide an example hook for `base.html` of both synchronous and asynchronous frontend asset loading.

## TODO

- [x] Have tests been written for the new code?
- [x] Has documentation been written/updated?
- [X] New dependencies (if any) added to requirements file
- [x] Add an entry to CHANGELOG.rst
- [x] Add yourself it AUTHORS.rst if you don't appear there

## Reviewer guidance

This PR also adds in JS Linting during the webpack build process, in order to compensate for the fact that the JSHint precommit hook doesn't work on Windows and had to be removed.

In addition, it simplifies the naming of frontend assets to provide a simpler naming convention, based solely on the name of their Python module in Kolibri, and the name of the Class in which they are defined in `kolibri_plugins.py`.

Finally, it provides a Javascript asynchronous frontend asset loader that can load both CSS and Javascript, in order to allow plugins to be registered to the frontend asynchronously and loaded when needed.

Adds a function for creating template tags with a hook to allow arbitrary plugins to plugin synchronously and asynchronously into any template.

Creates tags for these for base.html to allow for extension of base.html frontend functionality and as an example.

## Issues addressed

Fixes #41, and fixes #38 

## Documentation

Much inline documentation. Also updates frontend.rst.

### CODE REVIEW

**Comment by rtibbles** [comment]
Woops, broke coverage.

**Comment by rtibbles** [comment]
Woot, the asset-loader module has *full* code coverage!

File: .jshintrc [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (aronasorman) [inline .jshintrc:1]: Is it possible to put a comment with a link pointing to where these flags are defined?

File: assets/src/logging.js [hunk 2]
```diff
@@ -1,3 +1,4 @@
+'use strict';
```
Comment (aronasorman) [inline assets/src/logging.js:1]: If everything is already linted as strict, why do we need to still add this header?

Comment (rtibbles) [inline assets/src/logging.js:1]: `'use strict';` is an environment directive that is parsed by the browser itself. It will enforce the code to be executed in a strict context, which means that the Javascript engine will be less forgiving.

See here for details: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode

File: .jshintrc (hunk 1 above)
Comment (rtibbles) [inline .jshintrc:1]: Do we have one in travis.yml, or .gitignore? It's in the JSHint documentation.

**Comment by rtibbles** [comment]
Pushing latest changes in case anyone wants to look at them.

Tests have not been updated, so will break. Documentation needs to be written.

**Comment by rtibbles** [comment]
Some tests updated now, have also made a new tox_env for node, so that we now only run JS tests once!

**Comment by rtibbles** [comment]
@aronasorman This is now ready for final review.

**Comment by rtibbles** [comment]
gifrecord_2016-03-14_180602

Demonstration of callback buffering and arg passing in async loading mode!

**Comment by aronasorman** [comment]
Merge conflict.

File: kolibri/core/assets/src/mediator/mediator.js [hunk 3]
```diff
@@ -1,7 +1,127 @@
...
+    _execute_callback_buffer: function(plugin) {
+        if (typeof this._callback_buffer[plugin.name] !== 'undefined') {
+            _.forEach(this._callback_buffer[plugin.name], function(buffer) {
+                plugin[buffer.method].apply(plugin, buffer.args);
+            });
+            delete this._callback_buffer[plugin.name];
+        }
+    },
+
+    register_plugin_async: function(plugin_name, plugin_urls, events) {
```
Comment (aronasorman) [inline kolibri/core/assets/src/mediator/mediator.js]: `register_plugin_async` seems to have been defined twice. Is that intentional?

File: kolibri/core/assets/src/kolibri_core_app.js [hunk 4]
```diff
@@ -1,8 +1,32 @@
...
+var lib = {
+    Backbone: require('backbone'),
+    loglevel: logging,
+    _: require('lodash')
+};
+
+var Kolibri = _.extend({
+    lib: lib,
+    plugins: Mediator._plugin_registry
+}, _.pick(Mediator, public_methods));
```
Comment (aronasorman) [inline kolibri/core/assets/src/kolibri_core_app.js]: I'm not too familiar with `lodash`. What does this entire `_.extend` and `_.pick` snippet do?

File: webpack.config.js [hunk 5]
```diff
@@ -11,6 +11,18 @@ var js_source_dirs = [
...
 
-var bundles = recurseBundlePlugins(js_source_dirs, __dirname);
+// Mappings for libraries that we bundle in the Kolibri core app.
+
+var libs =  function (kolibri_name) {
+    return {
+        'lodash': kolibri_name + '.lib._',
+        'backbone': kolibri_name + '.lib.Backbone',
+        'loglevel': kolibri_name + '.lib.loglevel',
+        'kolibri': kolibri_name
```
Comment (aronasorman) [inline webpack.config.js:33]: So is this where the namespacing to `lib` happens? It seems like we return a dict in here, how exactly do these libraries then get namespaced when loading in the browser?

File: kolibri/plugins/example_plugin/assets/example/example_plugin.js [hunk 6]
```diff
@@ -1,5 +1,24 @@
...
-module.exports = {};
+var Plugin = require('plugin_base');
+
+var ExamplePlugin = Plugin.extend({
+
+    events: {
+        'something_happened': 'hello_world'
+    },
+    hello_world: function(plugin) {
+        if (plugin === this) {
```
Comment (aronasorman) [inline kolibri/plugins/example_plugin/assets/example/example_plugin.js]: Why is this `if` statement required? So this callback is not guaranteed to be called on this plugin's context?

**Comment by aronasorman** [comment]
@rtibbles can you point to where the whole event-callback machinery is implemented? I'd like to know when callbacks are triggered, and what the surounding context is (this is prompted after seeing that callbacks can be called outside of their plugin context.)

File: docs/dev/frontend.rst [hunk 7]
```diff
@@ -28,30 +28,47 @@ As such for a particular component of a frontend plugin, would appear in the ass
...
+        once: {
+            'nothing_happened': 'goodbye_world'
+        },
+
+        hello_world: function(message) {
+            logging.info('Hello world!', message);
+        },
+
+        goodbye_world: function(message) {
+            logging.info('Goodbye, cruel world!', message);
```
Comment (aronasorman) [inline docs/dev/frontend.rst:120]: Such is the life of a plugin.

**Comment by aronasorman** [comment]
Can you add a section in `frontend.rst` showing which libraries they have access to and don't have to download (a.k.a. `Kolibri.libs.*`)

File: kolibri/plugins/base.py [hunk 8]
```diff
@@ -111,23 +111,33 @@ def webpack_bundle_data(cls):
         :return: dict with keys "name", "entry_file", and, "external".
         "name" - is the module path that the frontend plugin has.
         "entry_file" - is the Javascript file that defines the plugin.
-        "external" - an optional flag used only by the kolibri_core plugin.
+        "external" - an optional flag currently used only by the core plugin.
+        "core" - an optional flag *only* ever used by the core plugin.
```
Comment (aronasorman) [inline kolibri/plugins/base.py:115]: Can you also document what the effects are of having the `core` flag?

**Comment by MCGallaspy** [comment]
Does `README.rst` need to be updated (namely the instructions for running the dev server)?

**Comment by MCGallaspy** [comment]
Also, which version of node do we support/recommend?

**Comment by MCGallaspy** [comment]
I was trying this out myself. This commit causes an error for me when running `npm run watch`: https://github.com/rtibbles/kolibri/commit/2f3129fd556882a29b64fbf479ad7b66014a73a8

This works fine, however:

```js
var onloadcss = require('fg-loadcss/src/onloadCSS');
/*
...
*/
module.exports = {
    assetLoader: assetLoader,
    onloadCSS: onloadcss
};
```

**Comment by rtibbles** [comment]
And does nothing, as `onloadCSS.js` fails to define a module.exports. The exports-loader that is causing the error for you allows the global variable defined in that file to be explicitly exported and used.

I assume you have run `npm install`, so I guess there may be an issue with the export-loader and windows? None reported here - https://github.com/webpack/exports-loader/issues

**Comment by rtibbles** [comment]
@MCGallaspy Seems to be an issue with node v5 (I have replicated with node v5 locally).

Let's pin our node at v4 for now.

**Comment by rtibbles** [comment]
Hrm, seems this is not true. Having cleaned out my node_modules dir, I am able to replicate this on all versions of node. Most curious.

**Comment by rtibbles** [comment]
OK, got the external loader working properly on Node v4 and Node v5.

**Comment by rtibbles** [comment]
@MCGallaspy README is already updated with how to run the npm scripts.

Should be agnostic to node version.

File: kolibri/core/assets/src/mediator/mediator.js (hunk 3 above)
Comment (rtibbles) [inline kolibri/core/assets/src/mediator/mediator.js]: I don't see it - are you confusing it with `register_plugin_sync`?

File: kolibri/core/assets/src/kolibri_core_app.js (hunk 4 above)
Comment (rtibbles) [inline kolibri/core/assets/src/kolibri_core_app.js]: `_.pick` takes certain attributes (the keys in `public_methods`) from the Mediator object and then `_.extend` adds them to the object that has the `lib` and `plugins` attributes defined.

File: webpack.config.js (hunk 5 above)
Comment (rtibbles) [inline webpack.config.js:33]: This is a convenience mapping to ensure that no plugins bundle in lodash, backbone, or loglevel if they do a direct require call on them. Instead, the reference will be converted into a reference to the copy that is on the core Kolibri object.

File: kolibri/plugins/example_plugin/assets/example/example_plugin.js (hunk 6 above)
Comment (rtibbles) [inline kolibri/plugins/example_plugin/assets/example/example_plugin.js]: Originally, I was triggering this from the 'plugin_registered' event, which also emits the plugin object as the first argument passed to a callback, this was just checking that the plugin in question was registered.

Ultimately, we can parse arbitrary data and arguments into callback functions from events.

File: kolibri/plugins/base.py (hunk 8 above)
Comment (rtibbles) [inline kolibri/plugins/base.py:115]: Wilco.

**Comment by rtibbles** [comment]
@aronasorman Note: All callbacks that are registered are executed in a way that ensures that their `this` in the execution context refers to the plugin that registered that callback. That way, every callback has a guaranteed execution context. See above for my explanation of the check that I made - it was not checking the value of `this` but rather checking the value of the event argument.

**Comment by MCGallaspy** [comment]
> Merge branch 'master' into frontend_asset_loading …

Personally I've come to prefer rebasing, since it results in cleaner commit histories. Of course, it does become a little problematic for very large or long-standing PRs...

**Comment by MCGallaspy** [comment]
> README is already updated with how to run the npm scripts.

I see that now, but it's under the "Testing" heading, which is a little odd. I'd expect to see it under "Running Kolibri", and it would be better still if I could run a dev server with one command instead of two.

File: kolibri/core/assets/src/mediator/mediator.js [hunk 9]
```diff
@@ -1,3 +1,241 @@
...
+ * multiple time and one time events. When all event listeners have been registered, any buffered callbacks are passed
+ * to the Plugin object, in case it was previously registered asynchronously.
+ */
+Mediator.prototype.register_plugin_sync = function(plugin) {
+
+    this._register_multiple_events(plugin);
+    this._register_one_time_events(plugin);
+
+    this._plugin_registry[plugin.name] = plugin;
+    this._clear_async_callbacks(plugin);
```
Comment (aronasorman) [inline kolibri/core/assets/src/mediator/mediator.js]: We should document that async callbacks are cleared after plugin loading.

File: kolibri/core/assets/src/mediator/mediator.js [hunk 10]
```diff
@@ -1,3 +1,241 @@
...
+    this._execute_callback_buffer(plugin);
```
Comment (aronasorman) [inline kolibri/core/assets/src/mediator/mediator.js]: Memory leaks are a concern. Not too much of a problem so early on, but let's watch out for this in the future.

**Comment by DXCanas** [comment]
Running into some weird issues on production server, discussed with Richard on 3/28. Traceback here:

`Environment:


Request Method: GET
Request URL: http://127.0.0.1:8000/

Django Version: 1.9.1
Python Version: 3.4.3
Installed Applications:
['kolibri.core',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'kolibri.auth.apps.KolibriAuthConfig',
 'kolibri.content']
Installed Middleware:
('django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware')



Traceback:

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in __getitem__
  86.             return self._engines[alias]

During handling of the above exception ('django'), another exception occurred:

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_package_libraries
  144.             module = import_module(entry[1])

File "/home/david/Envs/kolibri/lib/python3.4/importlib/__init__.py" in import_module
  109.     return _bootstrap._gcd_import(name[level:], package, level)

File "/home/david/Projects/kolibri/kolibri/core/templatetags/test/test_kolibri_tags.py" in <module>
  4. from mock.mock import patch

During handling of the above exception (No module named 'mock'), another exception occurred:

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/core/handlers/base.py" in get_response
  174.                     response = self.process_exception_by_middleware(e, request)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/core/handlers/base.py" in get_response
  172.                     response = response.render()

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in render
  160.             self.content = self.rendered_content

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in rendered_content
  135.         template = self._resolve_template(self.template_name)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in _resolve_template
  90.         new_template = self.resolve_template(template)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/response.py" in resolve_template
  80.             return select_template(template, using=self.using)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/loader.py" in select_template
  55.     engines = _engine_list(using)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/loader.py" in _engine_list
  143.     return engines.all() if using is None else [engines[using]]

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in all
  110.         return [self[alias] for alias in self]

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in <listcomp>
  110.         return [self[alias] for alias in self]

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/utils.py" in __getitem__
  101.             engine = engine_cls(params)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in __init__
  31.         options['libraries'] = self.get_templatetag_libraries(libraries)

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_templatetag_libraries
  49.         libraries = get_installed_libraries()

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_installed_libraries
  131.             for name in get_package_libraries(pkg):

File "/home/david/Envs/kolibri/lib/python3.4/site-packages/django/template/backends/django.py" in get_package_libraries
  148.                 "trying to load '%s': %s" % (entry[1], e)

Exception Type: InvalidTemplateLibrary at /
Exception Value: Invalid template library specified. ImportError raised when trying to load 'kolibri.core.templatetags.test.test_kolibri_tags': No module named 'mock'
`

**Comment by MCGallaspy** [comment]
I think with python 3, the `mock` module is actually `unittest.mock`. Looks like cause of the import error here in `test_kolibri_tags.py`. The fix is to use a conditional import like this:

```python
try:
    import mock  # Python 2
except ImportError:
    import unittest.mock as mock  # Python 3
```

**Comment by rtibbles** [comment]
No, this works fine in Python 3, as we have mock in the test requirements.
The issue is that Django is being more magic than I thought and importing
the test files as they've are inside the template tags folder. I will move
the tests out.

**Comment by MCGallaspy** [comment]
Oh, I see. Perhaps the tests directory should not be a module? I'm not really sure what the convention is here.

**Comment by rtibbles** [comment]
Two regression tests needed here: both for checking that `this` is set correctly. Will write before they are merged.

**Comment by rtibbles** [comment]
Have added appropriate regression tests. Fixed previous test failure.

Edited README.rst to move things out of testing that don't belong.

Fixed the testing issue for template tags.

Added line by line comments to some of the more obscure code.

File: .gitignore [hunk 11]
```diff
@@ -53,5 +54,5 @@ docs/modules.rst
 node_modules
 
 # webpack built files
-*stats.json
+kolibri/**/build/*
 kolibri/**/static/*
```
Comment (indirectlylit) [inline .gitignore:58]: I understand why this is necessary here, but the wildcard might cause brief confusion if someone decides to call a directory 'static' eventually

File: assets/src/parse_bundle_plugin.js [hunk 12]
```diff
@@ -8,14 +9,15 @@ var BundleTracker = require('webpack-bundle-tracker');
 var fs = require("fs");
 var path = require("path");
 var logging = require('./logging');
+var webpack = require('webpack');
+var base_config = require('../../webpack.config.base');
```
Comment (indirectlylit) [inline assets/src/parse_bundle_plugin.js:13]: relative path - code smell?

**Comment by rtibbles** [comment]
After some discussion with @indirectlylit and @MCGallaspy, and due to some issues with the approach taken to inspect the Javascript code and extract the `events` and `once` hashes at compilation time to allow for asynchronous loading - I have now changed the `events` and `once` hashes to be defined within the classes subclassed from `KolibriFrontEndBase`.

Configuration of these hashes now happens from within `kolibri_plugin.py`, and is then injected into the Javascript code at build time.

**Comment by MCGallaspy** [comment]
Can you also update this section: https://github.com/rtibbles/kolibri/blob/frontend_asset_loading/docs/dev/frontend.rst#writing-frontend-plugins

Also, who is responsible for merging this PR?

**Comment by rtibbles** [comment]
Good catch. Changing.

**Comment by rtibbles** [comment]
@aronasorman is assigned - I'll see if he is happy to sign it off.

File: assets/src/parse_bundle_plugin.js (hunk 12 above)
Comment (rtibbles) [inline assets/src/parse_bundle_plugin.js:13]: Trying to avoid relative paths as much as possible (hence the 'component' structure I have urged in the Frontend Docs - where code that will be imported together should be in the same directory) but in this case, I think a relative path is unavoidable unless we define an alias for node somewhere else.

File: kolibri/plugins/assets/src/plugin_base/plugin_base.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+    stop_listening: function(event, method) {
+        Kolibri.stop_listening(event, this, method);
+    }
+});
+
+// The Backbone Model extend method is a standalone function that is used to extend many Backbone objects.
+// We use it here in preference to rolling our own to allow for extension of Plugins.
+
+Plugin.extend = Backbone.Model.extend;
```
Comment (indirectlylit) [inline kolibri/plugins/assets/src/plugin_base/plugin_base.js]: Are plug-ins backbone models? If not, does this introduce a dependency on backbone just for this method?

Comment (indirectlylit) [inline kolibri/plugins/assets/src/plugin_base/plugin_base.js]: answering my own Q: Looks like `KolibriModule` is not a backbone model. We're also using `Backbone.Events` elsewhere.

Comment (indirectlylit) [inline kolibri/plugins/assets/src/plugin_base/plugin_base.js]: IMO it would be nice to inline the `extend` function to remove this instance of the Backbone dependency.

File: kolibri/core/kolibri_plugin.py [hunk 14]
```diff
@@ -38,6 +38,7 @@ class KolibriCoreFrontEnd(KolibriFrontEndPluginBase):
     """
     entry_file = "assets/src/kolibri_core_app.js"
     external = True
+    core = True
```
Comment (indirectlylit) [inline kolibri/core/kolibri_plugin.py:41]: Why is this added to `class KolibriCoreFrontEnd(KolibriFrontEndPluginBase)` but not `class KolibriCore(KolibriPluginBase)`?

Comment (indirectlylit) [inline kolibri/core/kolibri_plugin.py:41]: (Looks like `KolibriCore` is never used, and currently contains what looks like an example subclass implementation?)

**Comment by indirectlylit** [comment]
I'm a bit confused by the various webpack config files in the root directory:

* `parse_bundle_plugin.js` and `karma.conf.js` reference `webpack.config.base.js`
* `webpack.config.dev.js` references `webpack.config.js`
* `webpack.config.base.js` and `webpack.config.js` _do not_ reference each other

Seems like perhaps both `webpack.config.base.js` and `webpack.config.js` are trying to play the "base" role?

File: kolibri/plugins/assets/src/plugin_base/plugin_base.js (hunk 13 above)
Comment (rtibbles) [inline kolibri/plugins/assets/src/plugin_base/plugin_base.js]: Probably worthwhile.

File: kolibri/core/kolibri_plugin.py (hunk 14 above)
Comment (rtibbles) [inline kolibri/core/kolibri_plugin.py:41]: This 'core' property is only for defining a frontend KolibriModule as core. Should mention in the base class docs - if not, I'll add.

**Comment by rtibbles** [comment]
`webpack.config.base.js` is the base config that gets used for all other webpack configs (so shared across tests, production building, and dev building).

`parse_bundle_plugin.js` is at the end of a chain of three functions that build the configuration for `webpack.config.js`, so acts as the point of entry for the base config to get into the webpack configuration.

I will annotate the relevant files to make their roles more clear. Thanks for the feedback!

File: kolibri/core/templatetags/kolibri_tags.py [hunk 15]
```diff
@@ -36,15 +39,51 @@ def kolibri_main_navigation():
...
+    tags = [render_as_tags(get_webpack_bundle(callback(), None)) for callback in hooks.get_callables(hook)]
+    return mark_safe('\n'.join(tags))
+
+
+@register.simple_tag()
+def base_frontend_sync():
+    return frontend_sync(BASE_FRONTEND_SYNC)
+
+
+def frontend_async(hook):
```
Comment (indirectlylit) [inline kolibri/core/templatetags/kolibri_tags.py:137]: I'm till having a hard time wrapping my head around the design and use of the "hook" enumerated constants.

For example, `def frontend_async(hook)` and `def frontend_sync(hook)` aren't used anywhere except in this file. Also, I can't find where the "callbacks" are defined, and what they return.

Comment (indirectlylit) [inline kolibri/core/templatetags/kolibri_tags.py:137]: Ah, I see that subclasses can define a `hooks` method which returns a dict of functions. However these functions (e.g. `_register_front_end_plugins`, `main_navigation`) don't seem to have a consistent return value, but `get_webpack_bundle(bundle_path, extension)` seems to be called with the return value of the callbacks...

Comment (rtibbles) [inline kolibri/core/templatetags/kolibri_tags.py:137]: Yeah - this is where each hook has an 'implicit' API - based on function that its callbacks eventually get passed to.

I agree this could be improved - but it was like that when I got here.

Comment (indirectlylit) [inline kolibri/core/templatetags/kolibri_tags.py:137]: Above though we call `render_as_tags(get_webpack_bundle(callback(), None))` – doesn't that assume a particular return value from the callback?

Comment (rtibbles) [inline kolibri/core/templatetags/kolibri_tags.py:137]: Yes, that's what I mean by the implicit API of the hook. `_register_front_end_plugins` should be the method that is passed to the `FRONTEND_PLUGIN` hook as a callback. It will break if any other method is passed.

File: kolibri/core/templatetags/kolibri_tags.py [hunk 16]
```diff
@@ -36,15 +39,51 @@ def kolibri_main_navigation():
 
 def render_as_tags(bundle):
     tags = []
-    static = getattr(settings, 'STATIC_URL')
     for chunk in bundle:
-        url = chunk.get('publicPath') or chunk['url']
         if chunk['name'].endswith('.js'):
-            tags.append('<script type="text/javascript" src="{static}/{url}"></script>'.format(static=static, url=url))
+            tags.append('<script type="text/javascript" src="{url}"></script>'.format(url=render_as_url(chunk)))
```
Comment (indirectlylit) [inline kolibri/core/templatetags/kolibri_tags.py:50]: When module assets are loaded synchronously here (or asynchronously via `scriptjs`), it looks like they are executed in the global namespace. However files like `example_module.js` seem to leverage `module.exports` for namespacing. Is `module.exports` actually being used?

Comment (rtibbles) [inline kolibri/core/templatetags/kolibri_tags.py:50]: After webpack bundling, they are anonymous - the `module.exports` in `example_module.js` is vestigial and could be removed.

Comment (indirectlylit) [inline kolibri/core/templatetags/kolibri_tags.py:50]: @rtibbles clarified: the loaded JS is wrapped by `webpack` and not in the global namespace. However it is true that e.g. `module.exports = exampleModule;` has no effect, and all communication must happen through events with typical plugins rather than exported modules like in CommonJS and AMD.

**Comment by rtibbles** [comment]
This PR is now a month old. It is also being used in active code development by @MCGallaspy and @DXCanas - I believe all interested parties have had the opportunity to review, and I have responded to all comments.

@aronasorman has said to me that he does not have the time or familiarity to give the go ahead, as such, and due to the foregoing that there has been ample review, I will be merging this at EOD today if there are no strenuous objections raised prior to then.

**Comment by MCGallaspy** [comment]
Yeah, I think further iteration on this can be addressed in subsequent PRs.
